pkg/pool-weighted/contracts/test/MockManagedPoolSettings.sol

```diff
@@ -0,0 +1,123 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../managed/ManagedPoolSettings.sol";
+
+contract MockManagedPoolSettings is ManagedPoolSettings {
+    using WeightedPoolUserData for bytes;
+
+    constructor(
+        NewPoolParams memory params,
+        IVault vault,
+        IProtocolFeePercentagesProvider protocolFeeProvider,
+        address owner,
+        uint256 pauseWindowDuration,
+        uint256 bufferPeriodDuration
+    )
+        BasePool(
+            vault,
+            PoolRegistrationLib.registerPoolWithAssetManagers(
+                vault,
+                IVault.PoolSpecialization.MINIMAL_SWAP_INFO,
+                params.tokens,
+                params.assetManagers
+            ),
+            params.name,
+            params.symbol,
+            pauseWindowDuration,
+            bufferPeriodDuration,
+            owner
+        )
+        ManagedPoolSettings(params, protocolFeeProvider)
+    {
+        // solhint-disable-previous-line no-empty-blocks
+    }
+
+    function getVirtualSupply() external view returns (uint256) {
+        return _getVirtualSupply();
+    }
+
+    function _onInitializePool(address, bytes memory userData) internal override returns (uint256, uint256[] memory) {
+        WeightedPoolUserData.JoinKind kind = userData.joinKind();
+        _require(kind == WeightedPoolUserData.JoinKind.INIT, Errors.UNINITIALIZED);
+
+        (IERC20[] memory tokens, ) = _getPoolTokens();
+        uint256[] memory amountsIn = userData.initialAmountsIn();
+        InputHelpers.ensureInputLengthMatch(amountsIn.length, tokens.length);
+
+        uint256[] memory scalingFactors = _scalingFactors(tokens);
+        _upscaleArray(amountsIn, scalingFactors);
+
+        uint256 invariantAfterJoin = WeightedMath._calculateInvariant(_getNormalizedWeights(tokens), amountsIn);
+
+        // Set the initial BPT to the value of the invariant times the number of tokens. This makes BPT supply more
+        // consistent in Pools with similar compositions but different number of tokens.
+        uint256 bptAmountOut = Math.mul(invariantAfterJoin, amountsIn.length);
+
+        // We want to start collecting AUM fees from this point onwards. Prior to initialization the Pool holds no funds
+        // so naturally charges no AUM fees.
+        _lastAumFeeCollectionTimestamp = block.timestamp;
+
+        // amountsIn are amounts entering the Pool, so we round up.
+        _downscaleUpArray(amountsIn, scalingFactors);
+
+        return (bptAmountOut, amountsIn);
+    }
```

Comment:
Why do we need initialization?

Reply:
We perform some tests related to collecting AUM fees which requires the pool to be initialised.